File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* t1 = headA;
        int len1 = 0;
        while(t1 != NULL){
            len1++;
            t1 = t1->next;
        }

        ListNode* t2 = headB;
        int len2 = 0;
        while(t2 != NULL){
            len2++;
            t2 = t2->next;
        }
        ListNode* head1 = headA;
        ListNode* head2 = headB;
        int idx=0;
        if(len2>len1){
            idx = len2-len1;
            for(int i=1;i<=idx;i++){
                head2 = head2->next;
            }
        }
        else{
            idx = len1-len2;
            for(int i=1;i<=idx;i++){
                head1 = head1->next;
            }
        }
        bool flag  = false;
        while((head1 != NULL && head2 != NULL)){
            if(head1 != head2){
            head1 = head1->next;
            head2 = head2->next;
            }
            else{
                 flag = true;
                 break;
                 }
        }
        if(flag==true) return head1;
        else return NULL;
    }
};

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode l1 = headA;
        ListNode l2 = headB;
        int count1=0;
        int count2=0;
        while(l1 != null || l2 != null){
            if(l1 != null){
                count1++;
                l1 = l1.next;
            }
            if(l2 != null){
                count2++;
                l2 = l2.next;
            }
        }
        l1 = headA;
        l2 = headB;
        if(count2>count1){
            int k = count2-count1;
            for(int i=1;i<=k;i++){
                l2 = l2.next;
            }
        }
        if(count1>count2){
            int k = count1-count2;
            for(int i=1;i<=k;i++){
                l1 = l1.next;
            }
        }
        while(l2 != null && l1 != null){
            if(l1==l2) return l1;
            l1 = l1.next;
            l2 = l2.next;
        }
        return null;
    }
}